daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -41,15 +46,35 @@ class Hash private (private val bytes: Array[Byte]) {
 
 object Hash {
 
+  private val version = 0.toByte
+  private val underlyingHashLength = 32
+
+  implicit val HashOrdering: Ordering[Hash] =
+    ((hash1, hash2) => implicitly[Ordering[Iterable[Byte]]].compare(hash1.bytes, hash2.bytes))
+
   private[crypto] abstract class Builder {
 
-    def add(a: Array[Byte]): this.type
+    protected def update(a: Array[Byte]): Unit
 
-    def add(a: ByteBuffer): this.type
+    protected def doFinal(buf: Array[Byte], offset: Int): Unit
 
-    def add(a: Byte): this.type
+    final def build: Hash = {
+      val a = Array.ofDim[Byte](underlyingHashLength)
+      doFinal(a, 0)
+      new Hash(a)
+    }
 
-    def build: Hash
+    final def add(a: Array[Byte]): this.type = {
+      update(a)
+      this
+    }
+
+    private val byteBuff = Array.ofDim[Byte](1)
```

Comment:
this isn't thread-safe, but not sure if it's a problem...

Reply:
Builder are expected to be used internally of the `Hash` companion object. 
It is made package private for test purpose.